### PR TITLE
Prevent extensions from depending on extensions

### PR DESCRIPTION
**Pull request by jridgewell**
Maybe would have prevented #20392

### CODE REVIEW

File: build-system/dep-check-config.js [hunk 1]
```diff
@@ -351,6 +356,62 @@ exports.rules = [
   {
     filesMatching: 'src/**/*.js',
     mustNotDependOn: 'extensions/**/*.js',
+    whitelist: [
```
Comment (rsimha) [inline build-system/dep-check-config.js]: Shouldn't the whitelist be added to this section?

https://github.com/ampproject/amphtml/blob/87a2bdf6c97074585788d3592d8be6c81f61e580/build-system/dep-check-config.js#L207-L211

Comment (jridgewell) [reply build-system/dep-check-config.js]: Lolololol.

File: build-system/tasks/dep-check.js [hunk 2]
```diff
@@ -118,6 +118,17 @@ Rule.prototype.matchBadDeps = function(moduleName, deps) {
   deps.forEach(dep => {
     this.mustNotDependOn_.forEach(badDepPattern => {
       if (minimatch(dep, badDepPattern)) {
+
+        // Allow extension files to depend on their own code.
+        const dir = path.dirname(dep);
+        if (dir.startsWith('extensions/')) {
+          // eg, 'extensions/amp-geo/0.1'
+          const match = /extensions\/[^\/]+\/[^\/]+/.exec(dir);
```
Comment (rsimha) [inline build-system/tasks/dep-check.js]: Would this allow different versions of extensions to depend on each other? (Like `0.1` and `1.0`?) If so is it a problem?

File: build-system/dep-check-config.js [hunk 3]
```diff
@@ -204,6 +204,67 @@ exports.rules = [
   },
 
   // Rules for extensions.
+  {
+    // Extensions can't depend on other extensions.
+    filesMatching: 'extensions/**/*.js',
+    mustNotDependOn: 'extensions/**/*.js',
+    whitelist: [
```
Comment (rsimha) [inline build-system/dep-check-config.js]: Seems like this needs more entries: https://travis-ci.org/ampproject/amphtml/jobs/481109745#L716

**Comment by cramforce** [comment]
Wait, we didn't have this? Could it be that it was deleted at some point?

**Comment by jridgewell** [comment]
> Wait, we didn't have this? Could it be that it was deleted at some point?

I don't think we ever had it. Each extension would have to whitelist itself:

```text
extensions/amp-img/**/*.js->extensions/amp-img/**/*.js
extensions/amp-ad/**/*.js->extensions/amp-ad/**/*.js
```

With this new code, that's allowed by default. Only the exceptional cases need to be added to the whitelist.

**Pull request merged**